### DISCUSSION TITLE
package.json version gets reset to 0.0.0 when migrating an existing project

### DISCUSSION THREAD

**Discussion by janwvjaarsveld** [projen/projen#3193] [Q&A] [answered]
When migrating an existing typescript project to projen, the version inside the `package.json` gets reset to 0.0.0 every time I run `npx projen`. I can find a lot of information in the documentation about how to get started with a new project, but nowhere any information on how to migrate to projen and continue the project version from there. 

Can someone pleas point me in the correct direction of steps I need to follow for projen to pick up existing tags (and in what format they need to be) so that once I run `npx projen` it will set the version to the currently released version?

**Comment by mrgrain** [comment]
Hi! I think you are right to point out that the migration story of an existing project isn't very well documented at the moment.

I hope this documentation pages help you figure it out:
- [Why is the version in package.json set to 0.0.0?](https://projen.io/releases.html#why-is-the-version-in-packagejson-set-to-000)
- [How can I force a different version?](https://projen.io/releases.html#how-can-i-force-a-different-version)
- [Can I change the format of the release tag?](https://projen.io/releases.html#can-i-change-the-format-of-the-release-tag)
- [Custom self-set version of package?](https://github.com/projen/projen/discussions/2429)

**Reply by janwvjaarsveld** [reply]
Thank you for the quick reply! I am working through the pages to see if it solves my issue.

I do have a question on the way the versioning will then work for a monorepo scenario? How would projen know which tag belongs to which package, or would all packages need to be on the same version all the time?

Also with regards to [Cursom self-set version of package?](https://github.com/projen/projen/discussions/2429), would the version inside`addVersion()` need to be manually bumped every time you want an higher version?

**Reply by janwvjaarsveld** [reply]
And do I understand the following line correctly `To convey this message, the version in package.json is kept at 0.0.0.`, that the version in `package.json` will then always remain `0.0.0`, no matter the tagged version?

**Reply by mrgrain** [reply, marked as answer]
> Also with regards to https://github.com/projen/projen/discussions/2429, would the version insideaddVersion() need to be manually bumped every time you want an higher version?

Correct. If you want to manually manage the version, you will have to bump it manually. The automatic versioning works with `0.0.0` only. But since your concern is about migration, you might not want this anyway. Instead you can let it reset the version in `package.json` to `0.0.0` and follow [How can I force a different version?](https://projen.io/docs/publishing/releases-and-versioning/#how-can-i-force-a-different-version) to make sure projen does pickup the final non-projen version correctly.

> And do I understand the following line correctly To convey this message, the version in package.json is kept at 0.0.0., that the version in package.json will then always remain 0.0.0, no matter the tagged version?

Also correct.

**Reply by mrgrain** [reply]
> I do have a question on the way the versioning will then work for a monorepo scenario? How would projen know which tag belongs to which package, or would all packages need to be on the same version all the time?

You would probably use different tag prefixes per package. I am doing something similar here: https://github.com/cdklabs/awscdk-service-spec/tags

**Reply by janwvjaarsveld** [reply]
Thank you, I think that answers my questions then. And thanks for the repo, I will use it as an reference.

**Reply by adamelkhayyat** [reply]
> Correct. If you want to manually manage the version, you will have to bump it manually. The automatic versioning works with 0.0.0 only. But since your concern is about migration, you might not want this anyway. Instead you can let it reset the version in package.json to 0.0.0 and follow [How can I force a different version?](https://projen.io/releases.html#how-can-i-force-a-different-version) to make sure projen does pickup the final non-projen version correctly.

Doc update since the link in the answer is no longer working: https://projen.io/docs/publishing/releases-and-versioning/#how-can-i-force-a-different-version
